hazelcast-jet-kafka/src/main/java/com/hazelcast/jet/impl/connector/kafka/WriteKafkaP.java

```diff
@@ -52,28 +73,56 @@ public boolean isCooperative() {
     }
 
     @Override
-    @SuppressWarnings("unchecked")
-    protected boolean tryProcess(int ordinal, @Nonnull Object item) throws Exception {
-        producer.send(toRecordFn.apply((T) item));
-        return true;
+    public void process(int ordinal, @Nonnull Inbox inbox) {
+        checkError();
+        int drainedCount = inbox.drain((T item) -> {
+            // Note: send() method can block even though it is declared to not. This is true for Kafka 1.0 and probably
+            // will stay so, unless they change API.
+            producer.send(toRecordFn.apply(item), callback);
+        });
+        numPendingAsyncCalls.addAndGet(drainedCount);
+        producer.flush();
     }
 
     @Override
     public boolean complete() {
-        producer.flush();
+        ensureEverythingWritten();
         return true;
     }
 
+    @Override
+    public boolean saveToSnapshot() {
+        ensureEverythingWritten();
+        return true;
+    }
+
+    private void ensureEverythingWritten() {
+        checkError();
+        for (int idleCount = 0; numPendingAsyncCalls.get() > 0; idleCount++) {
+            IDLER.idle(idleCount);
+        }
+    }
+
+    private void checkError() {
+        Throwable t = lastError.get();
+        if (t != null) {
+            throw sneakyThrow(t);
+        }
+    }
+
     public static class Supplier<T, K, V> implements ProcessorSupplier {
 
-        static final long serialVersionUID = 1L;
+        private static final long serialVersionUID = 1L;
+        private static final String MAX_BLOCK_MS = Long.toString(1L << 32);
 
         private final Properties properties;
         private final Function<? super T, ProducerRecord<K, V>> toRecordFn;
 
         private transient KafkaProducer<K, V> producer;
 
         public Supplier(Properties properties, Function<? super T, ProducerRecord<K, V>> toRecordFn) {
+            // use a large enough value: we don't handle the TimeoutException and we don't want to fail due to it.
```

Comment:
I think it would be good to add this information(that we override `max.block.ms`) to the JavaDoc